### PR TITLE
Implement intitial iml-gui seed based framework.

### PR DESCRIPTION
**Pull request by jgrund**
This patch adds a new iml-gui crate that is written from the top-level
in seed.

It uses https://github.com/MartinKavik/seed-quickstart-webpack as it's
base and features:

  - Static css types.
  - Reload on save.
  - local dev server.

A skeleton of a refreshed GUI style is part of this patch.
Other components will need to be integrated / ported into this crate in
follow-on patches.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

### CODE REVIEW

File: iml-gui/configs/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        {
+          from: "static",
+          to: "static"
+        },
+        {
+          from: "favicons",
+          to: ""
+        },
+        {
+          from: "node_modules/@fortawesome/fontawesome-free/sprites",
```
Comment (johnsonw) [inline iml-gui/configs/webpack.config.js]: Is it `@fortawesome` or `@fontawesome`?

File: iml-gui/entries/templates/head_scripts.hbs [hunk 2]
```diff
@@ -0,0 +1,2 @@
+{{! Because of Edge, see https://github.com/samthor/fast-text-encoding }}
+<script type="text/javascript" src="/static/text-polyfill.min.js"></script>
```
Comment (johnsonw) [inline iml-gui/entries/templates/head_scripts.hbs]: Is this needed? We don't support IE still right?

File: iml-gui/favicons/site.webmanifest [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "",
```
Comment (johnsonw) [inline iml-gui/favicons/site.webmanifest]: No name?

File: iml-gui/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+- **Seed** - Rust framework, inspired by Elm. [Seed's Awesome list](https://github.com/MartinKavik/awesome-seed-rs).
+- **[Seed Quickstart](https://github.com/MartinKavik/seed-quickstart-webpack)** - This project is based on this quickstart template.
+- **[Tailwind CSS](https://tailwindcss.com/)** - CSS framework. All CSS classes in your project are typed for safe use in Rust code. Unused classes are automatically deleted for much smaller bundle size.
+- **[Webpack](https://webpack.js.org/)** - Bundler. Auto-reload on code change, dev-server accessible from mobile phones, prerendering for static websites... and many more useful features are prepared for you in this quickstart.
+
+## Dev Workflow
+
+1. Ensure global deps are installed:
+
+   - [Yarn](https://yarnpkg.com/lang/en/docs/install) - run `$ yarn -v` in terminal. It should output something like `1.17.3`
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-gui/README.md]: just what the quickstart came with. We can change it as it get's closer to production.

File: iml-gui/configs/webpack.config.js (hunk 1 above)
Comment (jgrund) [reply iml-gui/configs/webpack.config.js]: fortawesome.

File: iml-gui/entries/templates/head_scripts.hbs (hunk 2 above)
Comment (jgrund) [reply iml-gui/entries/templates/head_scripts.hbs]: For edge actually. Probably not, but we can always remove it later.

File: iml-gui/configs/postcss.config.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      : null;
+
+  return {
+    plugins: [
+      require("postcss-import"),
+      require("tailwindcss")("configs/tailwind.config.js"),
+      require("postcss-typed-css-classes")({
+        output_filepath: "crate/src/generated/css_classes.rs",
+        generator: "rust",
+        filter: class_ => {
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-gui/configs/postcss.config.js]: I'd rather not change this code as it came directly from the quickstart.

File: iml-gui/favicons/site.webmanifest (hunk 3 above)
Comment (jgrund) [reply iml-gui/favicons/site.webmanifest]: Updating.

File: iml-gui/crate/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+features = [
+  "Navigator",
+  "EventSource",
+  "DomRect",
+  "Element",
+  "MessageEvent"
+]
+
+[profile.release]
+lto = true
```
Comment (ip1981) [inline iml-gui/crate/Cargo.toml]: Does it make sense for wasm?

File: iml-gui/crate/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+repository = "https://github.com/whamcloud/integrated-manager-for-lustre"
+authors = ["IML Team <iml@whamcloud.com>"]
+description = "The IML GUI"
+categories = ["wasm", "rust", "tailwindcss", "seed"]
+license = "MIT"
+readme = "./README.md"
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib"]
```
Comment (ip1981) [inline iml-gui/crate/Cargo.toml]: Likewise, "cdylib" means "A dynamic system library". Does it make sense for wasm?

File: iml-gui/crate/src/generated/mod.rs [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// File `css_classes.rs` is (re)created during webpack compilation.
```
Comment (ip1981) [inline iml-gui/crate/src/generated/mod.rs]: Should this file be committed at all?

File: iml-gui/crate/src/page/partial/header.rs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+                    C.flex_grow,
+                    C.lg__flex,
+                    C.lg__items_center,
+                    C.lg__w_auto,
+                    C.lg__h_16,
+                ],
+                main_menu_items(&model),
+                div![a![
+                    attrs! {
+                        At::Class => "inline-block text-sm px-4 py-2 leading-none hover:text-teal-500 hover:bg-white mt-4 lg:mt-0",
```
Comment (ip1981) [inline iml-gui/crate/src/page/partial/header.rs]: Could it be `class![...]`?

File: iml-gui/scripts/postcss_rust_helpers.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+ * @returns usedCssClasses Set of class names
+ */
+function getUsedCssClassesInRust() {
+  const usedCssClasses = new Set();
+  // search in Rust files
+  const files = findFiles.fileSync(/\.rs$/, "./crate/src");
+  files.forEach(filePath => {
+    const fileContent = fs.readFileSync(filePath, "utf8");
+    // example of a used class in Rust code is `C.mb_16`
+    const usedCssClassesInFile = fileContent.match(/C\.[a-z0-9_]+/g) || [];
```
Comment (ip1981) [inline iml-gui/scripts/postcss_rust_helpers.js]: Will it pick up all `C.foo` even in comments, strings, etc.?

File: iml-gui/crate/Cargo.toml (hunk 6 above)
Comment (jgrund) [reply iml-gui/crate/Cargo.toml]: Yes, doing this tells llvm to optimize .wasm file further. Ref: https://rustwasm.github.io/book/reference/code-size.html#compiling-with-link-time-optimizations-lto

File: iml-gui/crate/Cargo.toml (hunk 7 above)
Comment (jgrund) [reply iml-gui/crate/Cargo.toml]: Yes, it's overloaded for Rust / wasm. Ref: https://rustwasm.github.io/wasm-pack/book/tutorials/npm-browser-packages/template-deep-dive/cargo-toml.html#1-crate-type

File: iml-gui/crate/src/generated/mod.rs (hunk 8 above)
Comment (jgrund) [reply iml-gui/crate/src/generated/mod.rs]: I think so. Without this file, we couldn't run `cargo check` on this crate without errors.

Likewise, it is generated by webpack and we probably shouldn't be required to run that every time for checking the crate.

File: iml-gui/scripts/postcss_rust_helpers.js (hunk 10 above)
Comment (jgrund) [reply iml-gui/scripts/postcss_rust_helpers.js]: Looks like it. Not sure if that's a problem in practice

File: iml-gui/crate/src/page/partial/header.rs (hunk 9 above)
Comment (jgrund) [reply iml-gui/crate/src/page/partial/header.rs]: Yes. This is a bit of cruft from initial conversion of the quickstart. I'll fix.

**Comment by ip1981** [comment]
In my Firefox 69.0.2  the slider is atop of the menu:

slider

**Comment by jgrund** [comment]
@ip1981 yes, still need to implement that submenu, but that will need to happen in a follow-on patch.

**Pull request merged**